Fix Mouseleave fires on elements no longer in the DOM

AlexsanderDamaceno opened the pull request:
#### 65d681444b66e8909e2e22e91d189c88d024674e
<pre>
Fix Mouseleave fires on elements no longer in the DOM
<a href="https://bugs.webkit.org/show_bug.cgi?id=156971">https://bugs.webkit.org/show_bug.cgi?id=156971</a>

Reviewed by NOBODY (OOPS!).

It fixes the bug where a deleted element fires  mouseleave events when we have mouse pointer inside it and move away from the element, even it is deleted.
by checking if the removed element node is still in the document tree. This way matching  the behaving from other browsers.

* LayoutTests/fast/events/mouse-events-on-node-deletion-expected.txt: Added.
* LayoutTests/fast/events/mouse-events-on-node-deletion.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateMouseEventTargetNode):
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_lostpointercapture_for_disconnected_node-expected.txt:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65d681444b66e8909e2e22e91d189c88d024674e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95978 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15592 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5538 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101039 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46487 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98023 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24024 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73169 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30395 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98981 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11883 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86690 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53505 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11629 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4442 "23 flakes 2 failures") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45822 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81783 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4543 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103066 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23045 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16788 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82211 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23296 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82713 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81584 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26174 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3610 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed layout tests; 12 flakes") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16382 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23008 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28163 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22667 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26147 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24408 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&button=0&buttons=1-expected.txt:
```diff
@@ -2,6 +2,7 @@
 FAIL Removing an element at mouseup assert_true: Removing an element at mouseup: Some events should've been fired after mousedown expected true got false
-PASS Removing an element at mousedown: mousedown should've been fired
-FAIL Removing an element at mouseup: mousedown should've been fired assert_equals: expected "{ 'mousedown' on 'child': button=0, buttons=1 }" but got "{ 'mousedown' on 'parent': button=0, buttons=1 }"
+FAIL Removing an element at mousedown: mousedown should've been fired assert_equals: expected "{ 'mousedown' on 'child': button=1, buttons=4 }" but got "{ 'mousedown' on 'child': button=1, buttons=2 }"
```
Why is it okay for this case to go from PASS to FAIL?

AlexsanderDamaceno replied:
@rniwa  I have fix this on my last commit, I think now the case is correct.

rniwa commented on LayoutTests/fast/events/mouse-events-on-node-deletion-expected.txt:
```diff
@@ -0,0 +1,20 @@
+On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE".
+
+
+outer received mouseover
```
This isn't the way js-test.js tests are typically written.
We should have a bunch of `shouldBe` function calls validating the outcome.
The current output doesn't make it clear when the test is passing
and when it's failing so we should update the test to do that.

rniwa commented on LayoutTests/fast/events/mouse-events-on-node-deletion.html:
```diff
@@ -0,0 +1,66 @@
+  eventSender.mouseMoveTo(0, 0);
+} else 
+{
+  debug("This test requires eventSender");
```
We should add a manual instruction for users to run this test within a browser.

AlexsanderDamaceno replied:
You mean that users should be able to run It manually?

rniwa replied:
Yes. See [LayoutTests/fast/events/remove-target-in-mouseup.html](https://github.com/WebKit/WebKit/blob/main/LayoutTests/fast/events/remove-target-in-mouseup.html) for an example.

AlexsanderDamaceno replied:
In this case I think is already  allowing the users execute manually, you can pass the mouse in the blocks and see the events happening.

![Screenshot from 2024-07-21 22-07-18](https://github.com/user-attachments/assets/3197baa1-33df-42ce-a8a7-1962eff3822e)

rniwa replied:
That doesn't tell us whether the test is passing or failing.

AlexsanderDamaceno replied:
@rniwa   I have rewritten the test case to use shouldbe calls, but I have not figured out a way to show a pass for the manual test. The pass would be shown when I delete the red block and move the mouse out of element space, and the inner mouse out event would not be fired as it is deleted from the document. However, I have not found a way to capture this moment to display the message, as I have already registered methods for the mouse event handler,.

rniwa commented on LayoutTests/fast/events/mouse-events-on-node-deletion.html:
```diff
@@ -0,0 +1,66 @@
+</style>
+
+<div id="outer" class="box">
+  <div id="inner" class="box"></div>
```
Please use 4-space indentation throughout the test.

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2782,12 +2782,16 @@ void EventHandler::updateMouseEventTargetNode(const AtomString& eventType, Node*
-            for (auto& chain : leftElementsChain) {
-                if (hasCapturingMouseLeaveListener || chain->hasEventListeners(eventNames.pointerleaveEvent) || chain->hasEventListeners(eventNames.mouseleaveEvent))
-                    chain->dispatchMouseEvent(platformMouseEvent, eventNames.mouseleaveEvent, 0, m_elementUnderMouse.get());
+            if (m_lastElementUnderMouse && m_lastElementUnderMouse->isConnected()) {
```
Why are we checking this condition again here?
We should just do this:
```
            if (auto lastElementUnderMouse = m_lastElementUnderMouse; lastElementUnderMouse && lastElementUnderMouse->isConnected()) {
                lastElementUnderMouse->dispatchMouseEvent(platformMouseEvent, eventNames.mouseoutEvent, 0, m_elementUnderMouse.get());                
                for (auto& chain : leftElementsChain) {
                    if (hasCapturingMouseLeaveListener || chain->hasEventListeners(eventNames.pointerleaveEvent) || chain->hasEventListeners(eventNames.mouseleaveEvent))
                        chain->dispatchMouseEvent(platformMouseEvent, eventNames.mouseleaveEvent, 0, m_elementUnderMouse.get());
                }
            }
```

AlexsanderDamaceno replied:
Ok, this is better, thank you for the test case update, I will make the changes.

rniwa commented on LayoutTests/platform/mac-wk1/fast/events/mouse-events-on-node-deletion-expected.txt:
```diff
@@ -0,0 +1,14 @@
+Verifies that mouse events are not fired on deleted nodes
```
Why do we need this file?

rniwa commented on LayoutTests/fast/events/mouse-events-on-node-deletion.html:
```diff
@@ -0,0 +1,86 @@
+    <!DOCTYPE HTML>
```
What's up with the indentation?
We don't normally indent the doc type like this.

AlexsanderDamaceno committed:
Fix Mouseleave fires on elements no longer in the DOM
https://bugs.webkit.org/show_bug.cgi?id=156971

Reviewed by NOBODY (OOPS!).

It fixes the bug where a deleted element fires  mouseleave events when we have mouse pointer inside it and move away from the element, even it is deleted.
by checking if the removed element node is still in the document tree. This way matching  the behaving from other browsers.

* LayoutTests/fast/events/mouse-events-on-node-deletion-expected.txt: Added.
* LayoutTests/fast/events/mouse-events-on-node-deletion.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateMouseEventTargetNode):
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_lostpointercapture_for_disconnected_node-expected.txt:

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2922,8 +2922,10 @@ void EventHandler::updateMouseEventTargetNode(const AtomString& eventType, Node*
             for (auto& chain : leftElementsChain) {
-                if (hasCapturingMouseLeaveListener || chain->hasEventListeners(eventNames.pointerleaveEvent) || chain->hasEventListeners(eventNames.mouseleaveEvent))
-                    chain->dispatchMouseEvent(platformMouseEvent, eventNames.mouseleaveEvent, 0, m_elementUnderMouse.get());
+                if (chain->isConnected()) {
```
What happens on other browsers when a removed node is inserted into another place in Document?

AlexsanderDamaceno replied:
You mean,  change element location  inside mouseleave event handling?

rniwa replied:
Right. One of these might be still connected to a document at this point in time if one of the event handlers had inserted it back into another location in the document. Heck, it could even re-insert itself back into the same location. We need to test that case to figure out what other browsers are doing.

AlexsanderDamaceno replied:
I wrote this test,  [https://jsfiddle.net/szau379o/5/](https://jsfiddle.net/szau379o/5/), removing the node inside  with mousemove and inserting in a new position.

would  like to check this behavior ?

AlexsanderDamaceno replied:
In Chrome is dispatching the mouseleave event but Firefox just dispatch the mousemove event.

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2922,8 +2922,10 @@ void EventHandler::updateMouseEventTargetNode(const AtomString& eventType, Node*
             for (auto& chain : leftElementsChain) {
-                if (hasCapturingMouseLeaveListener || chain->hasEventListeners(eventNames.pointerleaveEvent) || chain->hasEventListeners(eventNames.mouseleaveEvent))
-                    chain->dispatchMouseEvent(platformMouseEvent, eventNames.mouseleaveEvent, 0, m_elementUnderMouse.get());
+                if (chain->isConnected() || (!chain->isConnected() && chain->firstChild())) {
```
Why is an element with a child node special? That doesn't make much sense.

AlexsanderDamaceno replied:
This solves the problem when the child removed a parent that have a mouseleave event, this was only way found to aproach to this problem

rniwa replied:
I don't follow what you're trying to say. Please be more specific.
